internal/ethapi/api.go

```diff
@@ -921,6 +922,43 @@ func (diff *BlockOverrides) Apply(blockCtx *vm.BlockContext) {
 	}
 }
 
+type ChainContextBackend interface {
+	Engine() consensus.Engine
+	HeaderByNumber(context.Context, rpc.BlockNumber) (*types.Header, error)
+	HeaderByHash(context.Context, common.Hash) (*types.Header, error)
+}
+
+// ChainContext is an implementation of core.ChainContext for API backends
+// instead of BlockChain. Currently used by BlockChainAPI and traces.API.
+type ChainContext struct {
+	b   ChainContextBackend
+	ctx context.Context
+}
+
+// NewChainContext creates a new ChainContext object.
+func NewChainContext(ctx context.Context, backend ChainContextBackend) *ChainContext {
+	return &ChainContext{ctx: ctx, b: backend}
+}
+
+func (context *ChainContext) Engine() consensus.Engine {
+	return context.b.Engine()
+}
+
+func (context *ChainContext) GetHeader(hash common.Hash, number uint64) *types.Header {
```

Comment:
What's the intended semantics of this method? 
The underlying `HeaderByNumber` will always only return canonical headers. Do we want to support tracing on non-canon blocks? 

I haven't fully grokked why this `ChainContextBackend` is needed

Reply:
Aha then simply doing `HeaderByNumber` here should be enough. Good point we don't want non-canon blocks here because in the end this will be used by the `BLOCKHASH` opcode. And that's why I needed this `ChainContextBackend`, when constructing a new `vm.BlockContext` it takes in a `vm.ChainContext` as a parameter.

I'll update the method.